casper/src/test/scala/coop/rchain/casper/util/rholang/InterpreterUtilTest.scala

```diff
@@ -0,0 +1,154 @@
+package coop.rchain.casper.util.rholang
+
+import com.google.protobuf.ByteString
+
+import InterpreterUtil._
+import coop.rchain.casper.{BlockDag, BlockGenerator}
+import coop.rchain.casper.BlockDagState._
+import coop.rchain.casper.protocol._
+import coop.rchain.casper.util.ProtoUtil
+import org.scalatest.{FlatSpec, Matchers}
+
+import cats.Monad
+import cats.data.State
+import cats.implicits._
+import cats.mtl.implicits._
+
+import java.nio.file.Files
+
+import monix.execution.Scheduler.Implicits.global
+
+import scala.collection.immutable.HashMap
+
+class InterpreterUtilTest extends FlatSpec with Matchers with BlockGenerator {
+
+  type StateWithChain[A] = State[BlockDag, A]
+  val initState   = BlockDag().copy(currentId = -1)
+  val storageSize = 1024L * 1024
+
+  "computeBlockCheckpoint" should "compute the final post-state of a chain properly" in {
+    val storageDirectory = Files.createTempDirectory("casper-interp-util-test")
+
+    val genesisDeploys = Vector(
+      "@1!(1)",
+      "@2!(2)",
+      "for(@a <- @1){ @123!(5 * a) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b1Deploys = Vector(
+      "@1!(1)",
+      "for(@a <- @2){ @456!(5 * a) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b2Deploys = Vector(
+      "for(@a <- @123; @b <- @456){ @1!(a + b) }"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+
+    val b3Deploys = Vector(
+      "@7!(7)"
+    ).flatMap(mkTerm(_).toOption).map(ProtoUtil.termDeploy)
+    /*
+     * DAG Looks like this:
+     *
+     *          b3
+     *           |
+     *          b2
+     *           |
+     *          b1
+     *           |
+     *         genesis
+     */
+    def createChain[F[_]: Monad: BlockDagState]: F[BlockMessage] =
+      for {
+        genesis <- createBlock[F](Seq.empty, deploys = genesisDeploys)
+        b1      <- createBlock[F](Seq(genesis.blockHash), deploys = b1Deploys)
+        b2      <- createBlock[F](Seq(b1.blockHash), deploys = b2Deploys)
+        b3      <- createBlock[F](Seq(b2.blockHash), deploys = b3Deploys)
+      } yield b3
+    val chain   = createChain[StateWithChain].runS(initState).value
+    val genesis = chain.idToBlocks(0)
+
+    val b1 = chain.idToBlocks(1)
+    val b3 = chain.idToBlocks(3)
+
+    val (genCh, genMap) = computeBlockCheckpoint(genesis,
+                                                 genesis,
+                                                 chain,
+                                                 storageDirectory,
+                                                 storageSize,
+                                                 Map.empty[ByteString, Checkpoint])
+    val genPostState = genCh.toTuplespace.storageRepr
+    genPostState.contains("@{2}!(2)") should be(true)
+    genPostState.contains("@{123}!(5)") should be(true)
```

Comment:
Why you don't check whether it contains `"@1!(1)"` ?

Reply:
Because the only way `@123!(5)` could have appeared is if he `for`/`!` on `@1` reduced. If `@123` is there then I know that `@1` is not. I suppose it wouldn't hurt to add the extra check though